### PR TITLE
osd/PeeringState: set DELETING state and send to MPGStats

### PR DESCRIPTION
**Pull request by YiteGu**
When deleting a pool containing a large amount of data, OSD incurs
significant overhead and lasts for a long time. The pg deleting status
can easily let us know what osd is doing.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by YiteGu** [comment]
1. rm pool
```
ceph osd pool rm test-pool test-pool --yes-i-really-really-mean-it
```
2. we can get pg deleting state from ceph status
```
  cluster:
    id:     b5ef1685-9486-4d7c-9135-6a8846cf37a6
    health: HEALTH_WARN
            3 OSD(s) experiencing BlueFS spillover

  services:
    mon: 3 daemons, quorum a,b,c (age 2m)
    mgr: x(active, since 2m)
    mds: 1/1 daemons up, 2 standby
    osd: 3 osds: 3 up (since 113s), 3 in (since 2w)
    rgw: 1 daemon active (1 hosts, 1 zones)

  data:
    volumes: 1/1 healthy
    pools:   10 pools, 401 pgs
    objects: 5.07k objects, 19 MiB
    usage:   3.1 GiB used, 300 GiB / 303 GiB avail
    pgs:     7.980% pgs not active
             369 active+clean
             32  deleting
```
3. until the pg is delete completed
```
  cluster:
    id:     b5ef1685-9486-4d7c-9135-6a8846cf37a6
    health: HEALTH_WARN
            3 OSD(s) experiencing BlueFS spillover

  services:
    mon: 3 daemons, quorum a,b,c (age 2m)
    mgr: x(active, since 2m)
    mds: 1/1 daemons up, 2 standby
    osd: 3 osds: 3 up (since 2m), 3 in (since 2w)
    rgw: 1 daemon active (1 hosts, 1 zones)

  data:
    volumes: 1/1 healthy
    pools:   10 pools, 383 pgs
    objects: 2.44k objects, 9.2 MiB
    usage:   3.1 GiB used, 300 GiB / 303 GiB avail
    pgs:     3.655% pgs not active
             369 active+clean
             14  deleting
```
```
  cluster:
    id:     b5ef1685-9486-4d7c-9135-6a8846cf37a6
    health: HEALTH_WARN
            3 OSD(s) experiencing BlueFS spillover

  services:
    mon: 3 daemons, quorum a,b,c (age 2m)
    mgr: x(active, since 2m)
    mds: 1/1 daemons up, 2 standby
    osd: 3 osds: 3 up (since 2m), 3 in (since 2w)
    rgw: 1 daemon active (1 hosts, 1 zones)

  data:
    volumes: 1/1 healthy
    pools:   10 pools, 376 pgs
    objects: 1.37k objects, 5.0 MiB
    usage:   3.1 GiB used, 300 GiB / 303 GiB avail
    pgs:     1.862% pgs not active
             369 active+clean
             7   deleting
```
```
  cluster:
    id:     b5ef1685-9486-4d7c-9135-6a8846cf37a6
    health: HEALTH_WARN
            3 OSD(s) experiencing BlueFS spillover

  services:
    mon: 3 daemons, quorum a,b,c (age 2m)
    mgr: x(active, since 2m)
    mds: 1/1 daemons up, 2 standby
    osd: 3 osds: 3 up (since 2m), 3 in (since 2w)
    rgw: 1 daemon active (1 hosts, 1 zones)

  data:
    volumes: 1/1 healthy
    pools:   9 pools, 369 pgs
    objects: 261 objects, 716 KiB
    usage:   3.1 GiB used, 300 GiB / 303 GiB avail
    pgs:     369 active+clean
```

**Comment by YiteGu** [comment]
jenkins test api

**Comment by YiteGu** [comment]
API test fail seem to no related this PR. I think this is interesting features, please code reivew and discuss whether we need this feature. @neha-ojha

**Comment by Matan-B** [comment]
Reopening as this wasn't reviewed yet and it seems like an interesting enhancement

File: src/mgr/ClusterState.cc [hunk 1]
```diff
@@ -124,8 +113,11 @@ void ClusterState::ingest_pgstats(ref_t<MPGStats> stats)
 	       << q->second.reported_seq << dendl;
       continue;
     }
-
-    pending_inc.pg_stat_updates[pgid] = pg_stats;
+    if (pg_state_string(pg_stats.state) == "deleted") {
```
Comment (ronen-fr) [inline src/mgr/ClusterState.cc:116]: why compare against the string, instead of the actual state?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -7788,6 +7788,21 @@ MPGStats* OSD::collect_pg_stats()
 	min_last_epoch_clean_pgs.push_back(pg->pg_id.pgid);
       });
   }
+
+  {
+    std::lock_guard l(service.pg_delete);
+    auto& osd_deleting_pgs = service.get_osd_deleting_pgs();
+    for (auto& deleting_pg : osd_deleting_pgs) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7795]: const?

Comment (ronen-fr) [reply src/osd/OSD.cc:7795]: and maybe consider structured binding?

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -7788,6 +7788,21 @@ MPGStats* OSD::collect_pg_stats()
...
+      m->pg_stat[deleting_pg.first] = deleting_pg.second;
+    }
+    auto& osd_deleted_pgs = service.get_osd_deleted_pgs();
+    for (auto& deleted_pg : osd_deleted_pgs) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7799]: same here

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -7788,6 +7788,21 @@ MPGStats* OSD::collect_pg_stats()
...
+      m->pg_stat[deleted_pg.first] = deleted_pg.second;
+
+      // After sending it to MGR, it could be deleted.
+      osd_deleted_pgs.erase(deleted_pg.first);
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7803]: can't modify the container while traversing it!

File: src/osd/PG.cc [hunk 5]
```diff
@@ -2395,6 +2395,15 @@ std::pair<ghobject_t, bool> PG::do_delete_work(
     ++num;
   }
   bool running = true;
+
+  std::lock_guard l(osd->pg_delete);
```
Comment (ronen-fr) [inline src/osd/PG.cc:2399]: how frequent are the calls to collect_pg_stats()?
(i.e. - is there a cheaper place to execute this code?)

File: src/mgr/ClusterState.cc (hunk 1 above)
Comment (YiteGu) [reply src/mgr/ClusterState.cc:116]: If one pg will be delete, we no set pg state on `PeeringState::Deleting::Deleting(my_context ctx)`. The reason for comparing here is entirely to remove the pg in the previous deleting state. Perhaps we can provided a new method `pg_state_test` in `osd_types.h`

File: src/osd/PG.cc (hunk 5 above)
Comment (YiteGu) [reply src/osd/PG.cc:2399]: default 5s, controlled by option `mgr_stats_period`

File: src/osd/OSD.cc (hunk 2 above)
Comment (YiteGu) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 3 above)
Comment (YiteGu) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 4 above)
Comment (YiteGu) [reply src/osd/OSD.cc]: done

File: src/mgr/ClusterState.cc (hunk 1 above)
Comment (ronen-fr) [reply src/mgr/ClusterState.cc:116]: Let me rephrase the question:
why not
'if (pg_stats.state == <whatever the constant is for "deleted">) ...'

i.e. - why bother changing a constant to string?

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -7788,6 +7788,22 @@ MPGStats* OSD::collect_pg_stats()
 	min_last_epoch_clean_pgs.push_back(pg->pg_id.pgid);
       });
   }
+
+  {
+    std::lock_guard l(service.pg_delete);
+    auto& osd_deleting_pgs = service.get_osd_deleting_pgs();
+    for (auto& [pgid, pg_stat] : osd_deleting_pgs) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7795]: const auto&

File: src/osd/OSD.cc (hunk 3 above)
Comment (ronen-fr) [reply src/osd/OSD.cc]: const auto&

File: src/osd/OSD.cc [hunk 7]
```diff
@@ -7788,6 +7788,22 @@ MPGStats* OSD::collect_pg_stats()
...
+      m->pg_stat[pgid] = pg_stat;
+    }
+    auto& osd_deleted_pgs = service.get_osd_deleted_pgs();
+    if (!osd_deleted_pgs.empty()) {
+      for (auto& [pgid, pg_stat] : osd_deleted_pgs) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7800]: const

File: src/osd/PG.cc (hunk 5 above)
Comment (ronen-fr) [reply src/osd/PG.cc:2399]: @rzarzynski - can you review (esp this) as well? it might be negligible enough, or we may want to use something more complex here.

File: src/osd/OSD.cc (hunk 6 above)
Comment (YiteGu) [reply src/osd/OSD.cc]: done

File: src/osd/OSD.cc (hunk 7 above)
Comment (YiteGu) [reply src/osd/OSD.cc]: done

File: src/osd/PG.cc (hunk 5 above)
Comment (YiteGu) [reply src/osd/PG.cc]: @ronen-fr I will encapsulate the logic here into two functions, reduced the granularity of pg_delete lock, and the code is also more concise.

**Comment by YiteGu** [comment]
jenkins test make check
